Add HatsProposalCreationWhitelistV1 contract and associated interfaces

adamgall opened the pull request:
depends on https://github.com/decentdao/decent-contracts/pull/210

- Introduced `HatsProposalCreationWhitelistV1` contract for managing proposal creation permissions based on whitelisted Hats.
- Created `IHatsProposalCreationWhitelistV1` interface defining the contract's functionality and events.
- Developed `ConcreteHatsProposalCreationWhitelistV1` mock contract for testing purposes.
- Added `MockHats2` contract to simulate the Hats protocol for testing.
- Implemented comprehensive tests for `HatsProposalCreationWhitelistV1`, covering initialization, whitelisting, unwhitelisting, and proposal authorization logic.

twodam commented on contracts/interfaces/IHatsProposalCreationWhitelistV1.sol:
```diff
@@ -0,0 +1,72 @@
+     * @dev Adds a Hat to the whitelist for proposal creation.
+     * @param _hatId The ID of the Hat to whitelist
+     */
+    function whitelistHat(uint256 _hatId) external;
```
would `addHatToWhitelist` be more unify to style of `removeHatFromWhitelist`? or maybe use `whitelistHat` with `unwhitelistHat`?

adamgall replied:
yes I agree, will make this change

twodam commented on contracts/azorius/strategies/HatsProposalCreationWhitelistV1.sol:
```diff
@@ -0,0 +1,115 @@
+        (address _hatsContract, uint256[] memory _initialWhitelistedHats) = abi
+            .decode(initializeParams, (address, uint256[]));
+
+        if (_hatsContract == address(0)) revert InvalidHatsContract();
+        hatsContract = IHats(_hatsContract);
```
it can still be **invalid** hats contract even if it pass the check of zero address.

but I guess we has no reliable method to validate this? `!ERC165(_hatsContract).supportsInterface(type(IHats).interfaceId)` won't work because Hats contract itself didn't add that interfaceId to `supportsInterface`?

how about add something like this?
```solidity
(bool success, ) = _hatsContract.staticcall(
     abi.encodeWithSignature("isWearerOfHat(address,uint256)", address(this), 0)
);
require(success, "InvalidHatsContract");
```

adamgall replied:
although everything you're saying is true and accurate...

the zero address check is more about making sure that the parameter hasn't been accidentally omitted during the `setUp` call. considering that this `setUp` call is only made once per instance, and our own app is the one creating that calldata, I'm not overly concerned about this. i will probably not change this code.

adamgall replied:
however, I will change the error name from `InvalidHatsContract` to `MissingHatsContract` which better reflects the intention of this zero address check.

twodam commented on contracts/azorius/strategies/HatsProposalCreationWhitelistV1.sol:
```diff
@@ -0,0 +1,115 @@
+        bytes4 interfaceId
+    ) public view virtual override returns (bool) {
+        return
+            interfaceId == type(IHatsProposalCreationWhitelistV1).interfaceId ||
```
nice! I don't know why Hats didn't add interfaceId of their own `IHats` interface.

adamgall replied:
my assumption for why they didn't implement ERC165 in their contract: is because the Hats contract is a singleton instance onchain. there's already some assumption/expectation that if some outside caller is calling the Hats address, you know exactly what it is and therefore know what functions you can call on it.

but idk. i think it would be cleaner if they implemented ERC165

twodam commented on test/HatsProposalCreationWhitelistV1.test.ts:
```diff
@@ -0,0 +1,309 @@
+        deployer,
+      ).deploy();
+
+      // Create initialization data with zero address for hats contract
```
add test case where maybe someone accidentally pass a `ERC20` contract (or a contract guarantee to have no `isWearerOfHat` function) as hats contract? what's the consequence if someone did this? update to new strategy with correct hats contract?

adamgall replied:
consequence: the strategy is broken from the get-go
solution: deploy a new strategy and configure it correctly

again since this `setUp` call is a one time thing, and our own app is the one calling it, I'm not too concerned about this edge case ever happening in real life.

twodam commented on test/HatsProposalCreationWhitelistV1.test.ts:
```diff
@@ -0,0 +1,309 @@
+    });
+  });
+
+  describe('ERC165', () => {
```
As per the ERC-165 spec, no interface should ever match `0xffffffff`.

<img width="829" alt="image" src="https://github.com/user-attachments/assets/97e93416-05e6-44ba-9e0b-d111f8fbf626" />

adamgall replied:
I can add a test for this

adamgall commented:
@twodam i've implemented the changes i said i would, would appreciate a re-review when you get the chance. thank you!

adamgall committed:
Add HatsProposalCreationWhitelistV1 contract and associated interfaces

- Introduced `HatsProposalCreationWhitelistV1` contract for managing proposal creation permissions based on whitelisted Hats.
- Created `IHatsProposalCreationWhitelistV1` interface defining the contract's functionality and events.
- Developed `ConcreteHatsProposalCreationWhitelistV1` mock contract for testing purposes.
- Added `MockHats2` contract to simulate the Hats protocol for testing.
- Implemented comprehensive tests for `HatsProposalCreationWhitelistV1`, covering initialization, whitelisting, unwhitelisting, and proposal authorization logic.

dacebt commented on contracts/azorius/strategies/HatsProposalCreationWhitelistV1.sol:
```diff
@@ -0,0 +1,115 @@
+     * It iterates through all whitelisted Hat IDs and checks if the given address
+     * is wearing any of them using the Hats Protocol.
+     */
+    function isProposer(address _address) public view virtual returns (bool) {
+        for (uint256 i = 0; i < whitelistedHatIds.length; i++) {
+            if (hatsContract.isWearerOfHat(_address, whitelistedHatIds[i])) {
+                return true;
+            }
+        }
+        return false;
+    }
```
should we also be doing other checks here for standing of that hat? For example there are functions like:

https://github.com/Hats-Protocol/hats-protocol/blob/main/src/Hats.sol
- `checkHatWearerStatus` Line 372
- `isActive` Line 1042
- `isInGoodStanding` Line 1066

Or wearer being all we are looking for in this system?

adamgall replied:
- `checkHatWearerStatus` is not a view function, it can update state. so I don't want to call it in this `isProposer` function, which must be a view function
- `isWearerOfHat` calls `balanceOf` which checks both `isActive` and `isEligible`
- `isInGoodStanding` is interesting... I'm honestly not sure about this. Looking at the implementation, it calls a function on the given hat's `eligibility` module. We don't use `eligibility` modules in our system (afaik), so I don't think this is really relevant for us. I've only put a few minutes of thought into this at this point though.

dacebt replied:
- makes sense.

For the other two. yeah. I'm down to talk about this further.
